src/Middleware/RequestThrottling/src/IQueuePolicy.cs

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.RequestThrottling
+{
+    /// <summary>
+    /// Queueing policies, meant to be used with the <see cref="RequestThrottlingMiddleware"></see>.
+    /// </summary>
+    public interface IQueuePolicy
+    {
+        /// <summary>
+        /// Called for every incoming request.
+        /// When it returns 'true' the request procedes to the server.
+        /// When it returns 'false' the request is rejected immediately.
+        /// </summary>
+        Task<bool> TryEnterAsync();
+
+        /// <summary>
+        /// Called after successful requests have been returned from the server.
+        /// Does NOT get called for rejected requests.
+        /// </summary>
+        void OnExit();
```

Comment:
@davidfowl This is what I was talking to you about this afternoon. I think this is a good first step, but it doesn't allow the policy implementation track the time taken by individual requests. At best you can deduce the average time taken.

We talked about having TryEnterAsync return an interface that implements IDisposable instead. We also talked about queuing and dequeuing separately which would leave the middleware completely responsible for enforcing the concurrent request limit, though the policy could indirectly observe what the limit is.

The IDisposable idea doesn't seem too bad, but it's tricky to implement efficiently. The separate queue and dequeue seems like a major rearchitecture without any real concrete benefits other than it seems more pure and separates concerns better.

Reply:
After implementing some queueing policies, I will better understand what features the middleware needs, and can always go back to add them.

Having the middleware enforce the concurrent limit separately is a good separation of responsibility. However, the semaphore currently both enforces the limit and provides the queue.  Separating the responsibility would mean turning one semaphore into two, for the same end result.

Reply:
Yeah. This is perfectly fine for now. I just wanted more eyes on this in case there are some other good ideas we haven't considered yet.